topology/probes/sriov/sriov.go

```diff
@@ -0,0 +1,170 @@
+// +build linux
+
+/*
+ * Copyright (C) 2018 Samsung, Inc.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package sriov
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/vishvananda/netlink"
+
+	"github.com/skydive-project/skydive/common/fs"
+	"github.com/skydive-project/skydive/filters"
+	"github.com/skydive-project/skydive/logging"
+	"github.com/skydive-project/skydive/topology"
+	"github.com/skydive-project/skydive/topology/graph"
+)
+
+// Probe is a Sriov probe designed to handle all vfs on server
+type Probe struct {
+	graph   *graph.Graph
+	root    *graph.Node
+	cancel  context.CancelFunc // cancel function
+	watcher *fs.Watcher
+}
+
+// Manager is a const to be used as a string to describe who handles sriov nodes
+const Manager = "vfio"
+
+// GenerateNameForNodeFromPhysInterfaceNameAndVFID - generates a name for vf based on physical device name and id of of this vf
+func GenerateNameForNodeFromPhysInterfaceNameAndVFID(physName string, id string) string {
+	return fmt.Sprintf("%s-vf-%s", physName, id)
+}
+
+func allZero(s []byte) bool {
+	for _, v := range s {
+		if v != 0 {
+			return false
+		}
+	}
+	return true
+}
+
+func (h *Probe) getAllSriovNodes() []*graph.Node {
+	filter := filters.NewAndFilter(
+		filters.NewTermStringFilter("Manager", Manager),
+	)
+	m := graph.NewElementFilter(filter)
+	return h.graph.GetNodes(m)
+}
+
+func (h *Probe) periodicChecker(ctx context.Context) {
+	for ctx.Err() == nil {
+		linkList, err := netlink.LinkList()
+		if err != nil {
+			logging.GetLogger().Errorf("unable to get link list : %s", err)
+		}
+		h.graph.Lock()
+		var addedNodes = make(map[string]bool)
+		for _, link := range linkList {
+			attrs := link.Attrs()
+			parentNode := h.graph.LookupFirstChild(h.root, graph.Metadata{"Name": attrs.Name, "IfIndex": int64(attrs.Index)})
+			if parentNode == nil {
+				continue
+			}
+			for _, vf := range attrs.Vfs {
+				name := GenerateNameForNodeFromPhysInterfaceNameAndVFID(attrs.Name, strconv.Itoa(vf.ID))
+				if allZero(vf.Mac) {
```

Comment:
the call the allZero should probably be done before `GenerateNameForNodeFromPhysInterfaceNameAndVFID`

Reply:
done